File: go-implementation/internal/redis/publisher.go
package redis

import (
	"context"
	"encoding/json"
	"github.com/redis/go-redis/v9"
)

type Publisher struct {
	Client RedisAdapter
}

type RedisAdapter interface {
	Publish(ctx context.Context, channel string, message interface{}) *redis.IntCmd
	Close() error
}

func NewPublisher(client RedisAdapter) *Publisher {
	return &Publisher{Client: client}	
}

func (p *Publisher) Publish(channel string, message interface{}) error {
	jsonData, err := json.Marshal(message)

	if err != nil {
		return err
	}
	return p.Client.Publish(Ctx, channel, jsonData).Err()
}

func (p *Publisher) Close() error {
	return p.Client.Close()
}

File: node-implementation/src/core/Publisher.js
export class Publisher {
    constructor(redisClient) {
        this.client = redisClient;
    }

    async connect() {
        await this.client.connect();
    }

    async publish(channel, message) {
        await this.client.publish(channel, JSON.stringify(message));
    }

    async disconnect() {
        await this.client.quit();
    }
}